packages/inline-edit/src/InlineEdit.tsx

```diff
@@ -0,0 +1,158 @@
+import type { InlineVariants } from './styles/InlineEdit.css';
+import type { TextAreaProps, TextFieldProps } from '@launchpad-ui/form';
+import type {
+  ComponentProps,
+  Dispatch,
+  KeyboardEventHandler,
+  ReactElement,
+  SetStateAction,
+} from 'react';
+
+import { ButtonGroup, IconButton } from '@launchpad-ui/button';
+import { TextField } from '@launchpad-ui/form';
+import { Icon } from '@launchpad-ui/icons';
+import { useButton } from '@react-aria/button';
+import { focusSafely } from '@react-aria/focus';
+import { mergeProps, useUpdateEffect } from '@react-aria/utils';
+import { cx } from 'classix';
+import { cloneElement, useRef, useState } from 'react';
+
+import { container, cancelButton, inline, readButton } from './styles/InlineEdit.css';
+
+type InlineEditProps = ComponentProps<'div'> &
+  InlineVariants &
+  Pick<ComponentProps<'input'>, 'defaultValue'> & {
+    'data-test-id'?: string;
+    onConfirm: Dispatch<SetStateAction<string>>;
+    hideEdit?: boolean;
+    renderInput?: ReactElement<TextFieldProps | TextAreaProps>;
+    isEditing?: boolean;
+    onCancel?: () => void;
+    onEdit?: () => void;
+    cancelButtonLabel?: string;
+    editButtonLabel?: string;
+    confirmButtonLabel?: string;
+  };
+
+const InlineEdit = ({
+  'data-test-id': testId = 'inline-edit',
+  layout = 'horizontal',
+  children,
+  defaultValue,
+  onConfirm,
+  hideEdit = false,
+  renderInput = <TextField />,
+  'aria-label': ariaLabel,
+  isEditing: isEditingProp,
+  onCancel,
+  onEdit,
+  cancelButtonLabel = 'cancel',
+  editButtonLabel = 'edit',
+  confirmButtonLabel = 'confirm',
+}: InlineEditProps) => {
+  const [isEditing, setEditing] = useState(isEditingProp ?? false);
+  const inputRef = useRef<HTMLInputElement>(null);
+  const editRef = useRef<HTMLButtonElement>(null);
+  const controlled = isEditingProp !== undefined;
+
+  useUpdateEffect(() => {
+    if (controlled) {
+      setEditing(isEditingProp);
+    }
+  }, [isEditingProp]);
+
+  useUpdateEffect(() => {
+    isEditing
+      ? inputRef.current && focusSafely(inputRef.current)
+      : editRef.current && focusSafely(editRef.current);
+  }, [isEditing]);
+
+  const handleEdit = () => {
+    !controlled && setEditing(true);
+    onEdit?.();
+  };
+
+  const handleCancel = () => {
+    !controlled && setEditing(false);
+    onCancel?.();
+  };
+
+  const handleConfirm = () => {
+    onConfirm(inputRef.current?.value || '');
+    !controlled && setEditing(false);
+  };
+
+  const handleKeyDown: KeyboardEventHandler<HTMLInputElement> = (event) => {
```

Comment:
Was exploring the new stories in this branch locally, noticed that pressing `Esc` will cancel an in-progress edit which brought me here, wondering if we should also consider the case where a user blurs away from the input/cancel/save buttons while editing (either by clicking outside of the affected elements, or via keyboard navigation)?

There is some nuance to consider, but I'm wondering about the case where multiple inline-editable fields exist next to each other... how would a user navigate between them? For example: if I begin editing a field but then try to keyboard navigate to another editable field, what should happen? I'd expect the edit for the first field to be implicitly cancelled, but curious to hear your thoughts.

Reply:
Yeah no strong opinion on blur implicitly cancelling. Can try that out and see how that feels.

Reply:
> How does this overlap with routing?

Today we consider a form dirty if the input hasn't been saved, and dirty forms (in most places) cause route changes to be blocked and the user is forced to confirm the navigation... if we make this interaction modal, blurring a field that is being edited would discard the draft (and therefore not cause the field to be considered dirty).

Conversely, if this interaction isn't modal the following scenario seems possible: begin editing a field, change the field's value, then navigate to a different field... in that case the first field would need to be considered dirty to prevent unexpectedly losing edits.

I also agree that these questions aren't especially complicated, and that we don't necessarily have to answer them in this PR.

Reply:
That said, we should be opinionated here before we show up at UX crit.

My 2¢: Let's cancel on blur for now. That feels snappier to me. 😅

Reply:
Meeting with Antonio shortly and can go over this topic.

Reply:
Discussed with Antonio and we are gonna go with cancel on blur 👍